sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/DeltaMetricStorage.java

```diff
@@ -0,0 +1,138 @@
+      new ConcurrentHashMap<>();
+  private final List<DeltaAccumulation<T>> unreportedDeltas = new ArrayList<>();
+
+  DeltaMetricStorage(Aggregator<T> aggregator) {
+    this.aggregator = aggregator;
+  }
+
+  /**
+   * Allocates memory for a new metric stream, and returns a handle for synchronous recordings.
+   *
+   * @param attributes The identifying attributes for the metric stream.
+   * @return A handle that will (efficiently) record synchronous measurements.
+   */
+  public BoundStorageHandle bind(Attributes attributes) {
+    AggregatorHandle<T> aggregatorHandle = activeCollectionStorage.get(attributes);
+    if (aggregatorHandle != null && aggregatorHandle.acquire()) {
+      // At this moment it is guaranteed that the Bound is in the map and will not be removed.
+      return aggregatorHandle;
+    }
+
+    // Missing entry or no longer mapped, try to add a new entry.
+    aggregatorHandle = aggregator.createHandle();
+    while (true) {
+      AggregatorHandle<?> boundAggregatorHandle =
+          activeCollectionStorage.putIfAbsent(attributes, aggregatorHandle);
+      if (boundAggregatorHandle != null) {
+        if (boundAggregatorHandle.acquire()) {
+          // At this moment it is guaranteed that the Bound is in the map and will not be removed.
+          return boundAggregatorHandle;
+        }
+        // Try to remove the boundAggregator. This will race with the collect method, but only one
+        // will succeed.
+        activeCollectionStorage.remove(attributes, boundAggregatorHandle);
+        continue;
+      }
+      return aggregatorHandle;
+    }
+  }
+
+  /**
+   * Returns the latest delta accumulation for a specific collection handle.
+   *
+   * @param collector The current reader of metrics.
+   * @param collectors All possible readers of metrics.
+   * @return The delta accumulation of metrics since the last read of a the specified reader.
+   */
+  public synchronized Map<Attributes, T> collectFor(
+      CollectionHandle collector, Set<CollectionHandle> collectors) {
+    // First we force a collection
+    collectSynchronousDeltaAccumulationAndReset();
+    // Now build a delta result.
+    Map<Attributes, T> result = new HashMap<>();
+    for (DeltaAccumulation<T> point : unreportedDeltas) {
+      if (!point.wasReadBy(collector)) {
+        mergeInPlace(result, point.read(collector), aggregator);
+      }
+    }
+    // Now run a quick cleanup of deltas before returning.
+    unreportedDeltas.removeIf(delta -> delta.wasReadyByAll(collectors));
+    return result;
+  }
+
+  /**
+   * Collects the currently accumulated measurements from the concurrent-friendly synchronous
+   * storage.
+   *
+   * <p>All synchronous handles will be collected + reset during this method. Additionally cleanup
+   * related stale concurrent-map handles will occur. Any {@code null} measurements are ignored.
+   */
+  @GuardedBy("this")
```

Comment:
I think this is supposed to be on fields, not on methods (and errorprone will ensure the marked fields are only accessed under said lock)

Reply:
This did work for me in VisualStudio, but I can remove if you'd like.

Reply:
Working means if a not synchronized method accesses the ArrayList without taking the lock. Is vscode really able to understand that with this pattern? And is Gradle able to? The requirement is for Gradle to catch it, and if it can then I'm missing parts of the behavior of this annotation so happy to know that.

Reply:
By the way, think I'm just waiting on this one to merge

Reply:
Or we can merge after filing an issue about it if it seems hairy, since it's about thread safety want to make sure it doesn't get lost.

Reply:
Didn't realize this was still open (sorry soo many comments I was missing things).  Just went with synchronized for now and we can rework the class internals over time (there's a few TODOs for it anyway).